lib-js/spec-directory/spec-directory.ts

```diff
@@ -0,0 +1,143 @@
+  abstract listFiles(): Promise<string[]>
+  /** Returns true if this directory has the given filename as a subitem */
+  abstract hasFile(filename: string): boolean
+  /** The file contents of the given filename */
+  abstract readFile(filename: string): Promise<string>
+  /** Update the contents of the given file */
+  abstract writeFile(filename: string, contents: string): Promise<void>
+  /** Remove the file from this directory */
+  abstract removeFile(filename: string): Promise<void>
+
+  // Subdirectories
+
+  /** Get the list of subdrectory names of this directory */
+  abstract listSubdirs(): Promise<string[]>
+
+  /** Get the subdirectory at the provided path relative to this directory */
+  async atPath(path: string): Promise<SpecDirectory> {
+    if (!path) return this
+    const i = path.indexOf("/")
+    if (i === -1) {
+      return await this.subdir(path)
+    }
+    const child = await this.subdir(path.slice(0, i))
+    return await child.atPath(path.slice(i + 1))
+  }
+
+  // helper to get the subitem with the given name
+  protected abstract getSubdir(name: string): Promise<SpecDirectory>
+
+  /**
+   * Return the subitem of this directory corresponding to the given name
+   */
+  async subdir(name: string): Promise<SpecDirectory> {
+    // Cache the subitem so we always return the same one
+    if (!this.subitems[name]) {
+      this.subitems[name] = await this.getSubdir(name)
+    }
+    return this.subitems[name]
+  }
+
+  /** Return the list of subdirectories */
+  async subdirs(): Promise<SpecDirectory[]> {
+    const list = await this.listSubdirs()
+    return Promise.all(list.map((item) => this.subdir(item)))
+  }
+
+  // Spec Options
+
+  // Get the options from a physical options.yml file
+  async directOptions(): Promise<RunOptions> {
+    const contents = this.hasFile("options.yml")
+      ? await this.readFile("options.yml")
+      : ""
+    // TODO validate run options
+    return (yaml.safeLoad(contents) as RunOptions) || {}
+  }
+
+  /** Get the spec options of this directory, including those inherited from its parent */
+  async options() {
+    const opts = await this.directOptions()
+    return this.parentOpts ? mergeOptions(this.parentOpts, opts) : opts
+  }
+
+  // Test case info
+
+  /** Return whether this directory corresponds to a test case */
+  isTestDir() {
+    return this.hasFile("input.scss") || this.hasFile("input.sass")
+  }
+
+  // Iteration
+
+  async setup() {}
+  async cleanup() {}
+
+  private isMatch(paths: string[]) {
+    return paths.length === 0 || paths.some((path) => this.relPath() === path)
+  }
+
+  // TODO should we move this function out to its own file?
+  /**
+   * Iterate through the subpaths of this directory, running the iteratee
+   * on all
+   * @param paths the paths to match against
+   * @param iteratee the function to call for each matching subdirectory
+   */
+  async forEachTest(paths: string[], iteratee: SpecIteratee) {
```

Comment:
I decided to have two methods together because I wanted the file setup/cleanup to occur when each virtual directory is being iterated over. I know the ruby version does this the other way (compiles a list of test directories, then instantiates the files for each test directory) but I think this implementation is cleaner and requires less functionality to be implemented?

To replicate the old behavior we'd need a reference to the parent directory of virtual directories so that the files can be instantiated, and we need to read the input file to see if it imports anything so that we don't end up creating the whole archive for each file.

With the current implementation, each directory only needs to know about itself and its children, and only the archive root is responsible for knowing when to create and update files. IMO this is easier to think about (and implement and test). What do you think?